src/bin/pg_autoctl/pgsql.c

```diff
@@ -1279,21 +1277,32 @@ parsePgsrSyncStateAndWAL(void *ctx, PGresult *result)
 	}
 }
 
+
 /*
- * pgsql_get_wal_lag_from_standby queries a standby PostgreSQL server to get the
- * replication lag as seen from the pg_stat_wal_receiver system view.
+ * pgsql_get_received_lsn_from_standby queries a standby PostgreSQL server to get the
+ * received_lsn value from the pg_stat_wal_receiver system view.
+ *
+ * received_lsn is the latest lsn known to be received and flushed to the disk. It does
+ * not specify if it is applied or not. Caller should have allocated necessary memory
+ * for result value.
+ *
+ * We are collecting the latest WAL entry that is received successfully. It will be
+ * eventually applied to the receiving database.  This information will later be
```

Comment:
Add a note that explains that the data is known to be safe on the standby, and can be replayed later at end-of-recovery before promotion of the standby happens, please. That's the reasoning here, and it might be that we need to ensure that the WAL known to be received is actually replayed at promotion time, so please look into that (in Postgres documentation + source code if needed) and add your findings and references to the comment!

Reply:
added some more explanation

Reply:
I would feel better with some references (links) to Postgres documentation about the behaviour at promotion time where we can read that it will actually replay any WAL that has been received before promoting. Otherwise we have just implemented a data loss hasard. Also this is related to our choice of `synchronous_commit` default value, which is currently `on`, and might have to be `remote_apply` depending on your findings.

Reply:
https://www.postgresql.org/docs/12/warm-standby.html#STANDBY-SERVER-OPERATION states that

```
Standby mode is exited and the server switches to normal operation when pg_ctl promote is run or a trigger file is found (trigger_file). Before failover, any WAL immediately available in the archive or in pg_wal will be restored, but no attempt is made to connect to the master.
```

Reply:
The documentation that we are interested in seems to be at https://www.postgresql.org/docs/current/warm-standby.html#STANDBY-SERVER-OPERATION where the documentation sayth:

> Standby mode is exited and the server switches to normal operation when pg_ctl promote is run or a trigger file is found (trigger_file). Before failover, any WAL immediately available in the archive or in pg_wal will be restored, but no attempt is made to connect to the master.

Reply:
done